<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>netZooPy.sambar.sambar &mdash; netZooPy 0.9.1 documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../../index.html" class="icon icon-home"> netZooPy
            <img src="../../../_static/netZoo_logo_v14.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                0.9
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../install/index.html">Installation guide</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../functions/cli.html">Command Line Interface</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../functions/api.html">The Animals (API)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">Home</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../functions/index.html">Functions</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../tutos/index.html">Tutorials</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">netZooPy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../../index.html">Module code</a> &raquo;</li>
      <li>netZooPy.sambar.sambar</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for netZooPy.sambar.sambar</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>
<span class="kn">from</span> <span class="nn">scipy.spatial.distance</span> <span class="kn">import</span> <span class="n">pdist</span><span class="p">,</span><span class="n">cosine</span><span class="p">,</span><span class="n">squareform</span>
<span class="kn">from</span> <span class="nn">scipy.cluster.hierarchy</span> <span class="kn">import</span> <span class="n">linkage</span><span class="p">,</span><span class="n">cut_tree</span>
<span class="kn">import</span> <span class="nn">pkg_resources</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Description:</span>
<span class="sd">    Python implementation of the Subtyping Agglomerated Mutations By Annotation Relations (SAMBAR) method as implemented in R https://github.com/mararie/SAMBAR.</span>
<span class="sd">    SAMBAR, or Subtyping Agglomerated Mutations By Annotation Relations, is a method to identify subtypes based on somatic mutation data. SAMBAR was used to identify mutational subtypes in 23 cancer types from The Cancer Genome Atlas (Kuijjer ML, Paulson JN, Salzman P, Ding W, Quackenbush J, *British Journal of Cancer* (May 16, 2018), doi: 10.1038/s41416-018-0109-7, https://www.nature.com/articles/s41416-018-0109-7, BioRxiv, doi: https://doi.org/10.1101/228031).</span>

<span class="sd">Usage:</span>
<span class="sd">    To use the package you can import it using: ```import pysambar```. And then access the different functions implemented with ```pysambar.function()```.</span>
<span class="sd">    As an example you can find mutation data of Uterine Corpus Endometrial Carcinoma (UCEC) primary tumpor samples from The Cancer Genome Atlas. This data is in the ToyData folder as well as the MSigDb &quot;Hallmark&quot; gene sets. </span>
<span class="sd">    The program will compute the pathway mutation scores and clustering for *k*=2-4 (by default) and output the corrected mutation scores, the pathway mutation scores, and the clustering table. </span>

<span class="sd">Example:</span>
<span class="sd">    Run the SAMBAR method with the ToyData from the UCEC mutation data:</span>
<span class="sd">    ``` </span>
<span class="sd">    import pysambar as sm</span>
<span class="sd">    pathways, groups = sm.sambar(&quot;/ToyData/mut.ucec.csv&quot;,&quot;ToyData/esizef.csv&quot;,&#39;ToyData/genes.txt&#39;,&#39;ToyData/h.all.v6.1.symbols.gmt&#39;)</span>
<span class="sd">    ``` </span>
<span class="sd">    The output of this command will be four files:</span>
<span class="sd">    ``` pt_out.csv```  -&gt; Pathway mutation score matrix.</span>
<span class="sd">    ``` mt_out.csv```  -&gt; Processed gene mutation score matrix.</span>
<span class="sd">    ``` dist_matrix.csv ``` -&gt; Distance matrix with binomial distance in numpy condensed format.</span>
<span class="sd">    ``` clustergroups.csv```  -&gt; Matrix of pertinence to a cluster.</span>
<span class="sd">    The function also returns the pathway matrix dataframe and the cluster group dataframe as python variables.</span>

<span class="sd">Notes:</span>
<span class="sd">    Flags by default in the sambar function:</span>
<span class="sd">    ``` normPatient=True```  -&gt; Normalizes the mutation data by number of mutations in a sample.</span>
<span class="sd">    ``` kmin=2,kmax=4```  -&gt; Cut-offs of the cluster tree.</span>
<span class="sd">    ``` gmtMSigDB=True```  -&gt; If the signature file comes from MSigDB the second element of each line is removed to process the format available at the database. This can be toggled off if using a custom signature file.</span>
<span class="sd">    ``` subcangenes=True```  -&gt; Subset to cancer associated genes. By default uses the file provided in ToyData.</span>

<span class="sd">Functions:</span>
<span class="sd">            This package includes the functions ```sambar```,```desparsify```,```corgenelength```,```convertgmt```,```clustering``` as well as an implementation of the binomial distance (Millar dissimilarity from the package vegdist from R. To see the full description of each of this functions use ```help(pysambar.function)```.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="c1">## Default toydata files</span>
<span class="n">esize</span> <span class="o">=</span> <span class="n">pkg_resources</span><span class="o">.</span><span class="n">resource_filename</span><span class="p">(</span><span class="vm">__name__</span><span class="p">,</span> <span class="s1">&#39;esizef.csv&#39;</span><span class="p">)</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">pkg_resources</span><span class="o">.</span><span class="n">resource_filename</span><span class="p">(</span><span class="vm">__name__</span><span class="p">,</span> <span class="s1">&#39;genes.txt&#39;</span><span class="p">)</span>
<span class="n">sign</span>  <span class="o">=</span> <span class="n">pkg_resources</span><span class="o">.</span><span class="n">resource_filename</span><span class="p">(</span><span class="vm">__name__</span><span class="p">,</span> <span class="s1">&#39;h.all.v6.1.symbols.gmt&#39;</span><span class="p">)</span>
<span class="n">mut</span>   <span class="o">=</span> <span class="n">pkg_resources</span><span class="o">.</span><span class="n">resource_filename</span><span class="p">(</span><span class="vm">__name__</span><span class="p">,</span> <span class="s1">&#39;mut.ucec.csv&#39;</span><span class="p">)</span>

<div class="viewcode-block" id="corgenelength"><a class="viewcode-back" href="../../../functions/api.html#netZooPy.sambar.corgenelength">[docs]</a><span class="k">def</span> <span class="nf">corgenelength</span><span class="p">(</span><span class="n">mut</span><span class="p">,</span><span class="n">cangenes</span><span class="p">,</span><span class="n">esize</span><span class="p">,</span><span class="n">normbysample</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">subcangenes</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Description:</span>
<span class="sd">        Function to normalize gene mutation scores by gene length.</span>
<span class="sd">        mut should be a dataframe of mutation scores with genes as columns and samples as rows. </span>
<span class="sd">        (VERY IMPORTANT, IF OTHERWISE MATRIX SHOULD BE TRANSPOSED OR IT WON&#39;T WORK!!!)</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    --------------</span>
<span class="sd">        mut         : Mutation scores.</span>
<span class="sd">        cangenes    : A set of cancer associated genes.</span>
<span class="sd">        esize       : A dataframe of gene lengths.</span>
<span class="sd">        normbysample: True : Normalizes the gene mutation scores in a sample by the total mutations within the sample. </span>
<span class="sd">                      False: Deactivate normalization.</span>
<span class="sd">        subcangenes : True: Subsets mutation data to cancer-associated genes.</span>
<span class="sd">                      False: Takes all genes.</span>

<span class="sd">    Returns	</span>
<span class="sd">    --------------</span>
<span class="sd">        mut:_         </span>
<span class="sd">            Mutation scores normalized by gene length.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#Subsets mutation data to cancer-associated genes.</span>
    <span class="k">if</span><span class="p">(</span><span class="n">subcangenes</span><span class="p">):</span>
    	<span class="n">mut_cangenes</span> <span class="o">=</span> <span class="n">cangenes</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">mut</span><span class="o">.</span><span class="n">columns</span><span class="p">))</span>
    	<span class="n">mut</span> <span class="o">=</span> <span class="n">mut</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">mut_cangenes</span><span class="p">)]</span>
    <span class="k">else</span><span class="p">:</span> <span class="n">mut_cangenes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">mut</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
    <span class="c1">#Intersection of genes in the mutation data and the length data.</span>
    <span class="n">mut_esize</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">mut_cangenes</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">esize</span><span class="o">.</span><span class="n">columns</span><span class="p">)))</span>
    <span class="n">mut</span> <span class="o">=</span> <span class="n">mut</span><span class="p">[</span><span class="n">mut_esize</span><span class="p">]</span>
    <span class="n">esize</span> <span class="o">=</span> <span class="n">esize</span><span class="p">[</span><span class="n">mut_esize</span><span class="p">]</span>
    
    <span class="c1">#Divide each number of mutations by gene size</span>
    <span class="n">mut</span> <span class="o">=</span> <span class="n">mut</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">esize</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    
    <span class="c1">#Drop samples (rows) without mutations (each column equals 0)</span>
    <span class="n">mut</span> <span class="o">=</span> <span class="n">mut</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">mut</span><span class="p">[(</span><span class="n">mut</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="mi">1</span><span class="p">)]</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
    
    <span class="c1">#Normalize mutation score by patient number of mutations</span>
    <span class="k">if</span><span class="p">(</span><span class="n">normbysample</span><span class="p">):</span>
        <span class="n">mut</span><span class="p">[</span><span class="s2">&quot;sum&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">mut</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="c1">#Adds auxiliar column with the sum of columns (sum of mutation scores within a sample)</span>
        <span class="n">mut</span> <span class="o">=</span> <span class="n">mut</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">mut</span><span class="p">[</span><span class="s2">&quot;sum&quot;</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="c1">#Corrects gene mutation scores in a sample by the total mutation score in the sample.</span>
        <span class="n">mut</span> <span class="o">=</span> <span class="n">mut</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sum&quot;</span><span class="p">])</span>
    
    <span class="c1">#Returns mut matrix of gene mutation scores by sample adjusted by length and total number of mutations within sample.</span>
    <span class="c1">#Sorted columns for niceness</span>
    <span class="k">assert</span> <span class="n">mut</span><span class="o">.</span><span class="n">shape</span><span class="o">!=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span><span class="s2">&quot;Are you sure your mutation matrix has the proper orientation? Genes as columns, samples as rows.&quot;</span>
    <span class="k">return</span> <span class="n">mut</span><span class="p">[</span><span class="nb">sorted</span><span class="p">(</span><span class="n">mut</span><span class="o">.</span><span class="n">columns</span><span class="p">)]</span></div>

<div class="viewcode-block" id="convertgmt"><a class="viewcode-back" href="../../../functions/api.html#netZooPy.sambar.convertgmt">[docs]</a><span class="k">def</span> <span class="nf">convertgmt</span><span class="p">(</span><span class="n">gmtfile</span><span class="p">,</span> <span class="n">cangenes</span><span class="p">,</span><span class="n">gmtMSigDB</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">subcangenes</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>

    <span class="sd">&quot;&quot;&quot; This function takes as input the name of a gmt file containing lists of genes associated to pathways. </span>
<span class="sd">        It outputs an adjacency matrix of genes and pathways. </span>
<span class="sd">        It also subsets the genes to a list of cancer-associated genes. </span>

<span class="sd">    Args:</span>
<span class="sd">        gmtfile (str): Path the gmt file.</span>
<span class="sd">        cangenes (list): A set of cancer associated genes.</span>
<span class="sd">        gmtMSigDB (bool, optional): If true: gmt file from MSigDB . Defaults to True.</span>
<span class="sd">        subcangenes (bool, optional): If true: Subsets mutation data to cancer-associated genes.</span>
<span class="sd">        otherwise takes all genes. Defaults to True.</span>

<span class="sd">    Returns:</span>
<span class="sd">        sign_matrix: _</span>
<span class="sd">            Adjacency matrix of genes and pathways.</span>
<span class="sd">    &quot;&quot;&quot;</span>    

    <span class="n">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">gmtfile</span><span class="p">)</span> <span class="c1"># Loads the gmt file</span>
    <span class="n">pw</span> <span class="o">=</span> <span class="n">file</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span> <span class="c1">#Reads file into list of strings</span>

    <span class="n">a</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span> <span class="c1">#Sets up dictionary</span>
    <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">pw</span><span class="p">:</span>
        <span class="n">u</span> <span class="o">=</span> <span class="n">entry</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="c1">#Process each string separating genes and removing final \n.</span>
        <span class="k">if</span><span class="p">(</span><span class="n">gmtMSigDB</span><span class="p">):</span>
        	<span class="n">a</span><span class="p">[</span><span class="n">u</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">u</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span> <span class="c1">#Sets dictionary key the pathway id, value the list of genes removing the link entry.</span>
        <span class="k">else</span><span class="p">:</span> <span class="n">a</span><span class="p">[</span><span class="n">u</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">u</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
    <span class="c1">#Set of all genes in the dictionary</span>
    <span class="n">b</span><span class="o">=</span><span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">a</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">b</span> <span class="o">=</span> <span class="n">b</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">a</span><span class="p">[</span><span class="n">i</span><span class="p">]))</span>

    <span class="n">C</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">b</span><span class="p">))</span> <span class="c1">#List of genes in the file (sorted)</span>
    <span class="n">P</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="c1">#List of pathways in the file (sorted)</span>
    
    <span class="n">G</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">(</span><span class="n">a</span><span class="p">)</span> <span class="c1">#Creates a bipartite graph from the dictionary</span>
    <span class="n">A</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">adjacency_matrix</span><span class="p">(</span><span class="n">G</span><span class="p">,</span><span class="n">nodelist</span><span class="o">=</span><span class="p">(</span><span class="n">C</span><span class="o">+</span><span class="n">P</span><span class="p">))</span><span class="o">.</span><span class="n">todense</span><span class="p">()</span> <span class="c1">#Creates the adjacency matrix of the graph G in the proper order. </span>
    <span class="c1">#Note that this matrix is square with size len(C)+len(P), the biadjacency matrix caused issues with big files.</span>
    
    <span class="c1">#Builds a dataframe from the adjacency matrix and tags the rows and columns.</span>
    <span class="n">sign_matrix</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">A</span><span class="p">)</span>
    <span class="n">sign_matrix</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="n">C</span><span class="o">+</span><span class="n">P</span>
    <span class="n">sign_matrix</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">C</span><span class="o">+</span><span class="n">P</span>
    
    <span class="c1">#Subsets the dataframe so it has as columns the genes and as rows the pathways. Orders for niceness.</span>
    <span class="n">sign_matrix</span> <span class="o">=</span> <span class="n">sign_matrix</span><span class="p">[</span><span class="nb">sorted</span><span class="p">(</span><span class="n">C</span><span class="p">)]</span>
    <span class="n">sign_matrix</span> <span class="o">=</span> <span class="n">sign_matrix</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="nb">sorted</span><span class="p">(</span><span class="n">P</span><span class="p">)]</span>
 
    <span class="c1">#Subset pathways to cancer-associated genes</span>
    <span class="k">if</span><span class="p">(</span><span class="n">subcangenes</span><span class="p">):</span>
    	<span class="n">sign_cangenes</span> <span class="o">=</span> <span class="n">cangenes</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">sign_matrix</span><span class="o">.</span><span class="n">columns</span><span class="p">))</span>
    	<span class="n">sign_matrix</span> <span class="o">=</span> <span class="n">sign_matrix</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">sign_cangenes</span><span class="p">)]</span>
    
    <span class="k">return</span> <span class="n">sign_matrix</span><span class="p">[</span><span class="nb">sorted</span><span class="p">(</span><span class="n">sign_matrix</span><span class="o">.</span><span class="n">columns</span><span class="p">)]</span></div>

<div class="viewcode-block" id="desparsify"><a class="viewcode-back" href="../../../functions/api.html#netZooPy.sambar.desparsify">[docs]</a><span class="k">def</span> <span class="nf">desparsify</span><span class="p">(</span><span class="n">mutdata</span><span class="p">,</span><span class="n">exonsize</span><span class="p">,</span><span class="n">gmtfile</span><span class="p">,</span><span class="n">cangenes</span><span class="p">,</span><span class="n">normMut</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">gmtMSigDB</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">subcangenes</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Applies the sambar method to de-sparcify the mutation data using the pathway signatures in the gmtfile.</span>
<span class="sd">        </span>
<span class="sd">    Args:</span>
<span class="sd">        mutdata     : Mutation scores.</span>
<span class="sd">        exonsize    : A dataframe of gene lengths.</span>
<span class="sd">        gmtfile     : Path the gmt file.</span>
<span class="sd">        cangenes    : A set of cancer associated genes.</span>
<span class="sd">        normbysample: True : Normalizes the gene mutation scores in a sample by the total mutations within the sample. </span>
<span class="sd">                      False: Deactivate normalization.</span>
<span class="sd">        gmtMSigDB   : True : gmt file from MSigDB </span>
<span class="sd">                      False: file not form MSigDB</span>
<span class="sd">        subcangenes : True : Subsets mutation data to cancer-associated genes.</span>
<span class="sd">                      False: Takes all genes.</span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    mt : _</span>
<span class="sd">        Genes in both pathways and mutation data</span>
<span class="sd">    pathway_scores : _</span>
<span class="sd">        Pathway scores.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">tinit</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>  
    
    <span class="c1">#Gets corrected gene mutation scores. Subseted to cancer-associated genes.</span>
    <span class="n">mutrate</span> <span class="o">=</span> <span class="n">corgenelength</span><span class="p">(</span><span class="n">mutdata</span><span class="p">,</span><span class="n">cangenes</span><span class="p">,</span><span class="n">exonsize</span><span class="p">,</span><span class="n">normMut</span><span class="p">,</span><span class="n">subcangenes</span><span class="p">)</span>
    <span class="c1">#Gets adjacency matrix of genes and pathways. Subseted to cancer-associated genes.</span>
    <span class="n">sign_matrix</span> <span class="o">=</span> <span class="n">convertgmt</span><span class="p">(</span><span class="n">gmtfile</span><span class="p">,</span><span class="n">cangenes</span><span class="p">,</span><span class="n">gmtMSigDB</span><span class="p">,</span><span class="n">subcangenes</span><span class="p">)</span>

    <span class="c1">#Match genes in both lists</span>
    <span class="n">genes_ms</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">mutrate</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">sign_matrix</span><span class="o">.</span><span class="n">columns</span><span class="p">))))</span>
    <span class="c1">## Computes the number of genes in a pathway (before matching with mutation data).</span>
    <span class="n">sm</span> <span class="o">=</span> <span class="n">sign_matrix</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    
    <span class="c1">#Matches genes in mutation data with genes in pathway data</span>
    <span class="n">mutrate</span> <span class="o">=</span> <span class="n">mutrate</span><span class="p">[</span><span class="n">genes_ms</span><span class="p">]</span>
    <span class="n">sign_matrix</span> <span class="o">=</span> <span class="n">sign_matrix</span><span class="p">[</span><span class="n">genes_ms</span><span class="p">]</span>
    <span class="n">mt</span> <span class="o">=</span> <span class="n">mutrate</span>
    <span class="c1">#Corrects mutation rates by frequency in a pathway</span>
    <span class="n">genefreq</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">sign_matrix</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span>
    <span class="n">mutrate</span> <span class="o">=</span> <span class="n">mutrate</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">genefreq</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
    
    <span class="c1">#This should be done here, otherwise the matching before checkpoint 2 might generate rows with only 0. SHOULD NOT BE COMMENTED</span>
    <span class="n">sign_matrix</span> <span class="o">=</span> <span class="n">sign_matrix</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">sign_matrix</span><span class="p">[(</span><span class="n">sign_matrix</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="mi">1</span><span class="p">)]</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
    <span class="n">sign_matrix</span> <span class="o">=</span> <span class="n">sign_matrix</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
    <span class="n">sign_matrix</span> <span class="o">=</span> <span class="n">sign_matrix</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">sign_matrix</span><span class="p">[(</span><span class="n">sign_matrix</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="mi">1</span><span class="p">)]</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
    <span class="n">sign_matrix</span> <span class="o">=</span> <span class="n">sign_matrix</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
    
    <span class="c1">#Matrix multiply dataframes to get the pathway mutation scores</span>
    <span class="n">pathway_scores</span> <span class="o">=</span> <span class="n">sign_matrix</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">mutrate</span><span class="p">)</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">sm</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">pathway_scores</span> <span class="o">=</span> <span class="n">pathway_scores</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span> <span class="c1">#Removes pathways with NaN values.</span>
    <span class="c1">#Remove rows and columns with only 0.</span>
    <span class="n">pathway_scores</span> <span class="o">=</span> <span class="n">pathway_scores</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">pathway_scores</span><span class="p">[(</span><span class="n">pathway_scores</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="mi">1</span><span class="p">)]</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
    <span class="n">pathway_scores</span> <span class="o">=</span> <span class="n">pathway_scores</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
    <span class="n">pathway_scores</span> <span class="o">=</span> <span class="n">pathway_scores</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">pathway_scores</span><span class="p">[(</span><span class="n">pathway_scores</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="mi">1</span><span class="p">)]</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
    <span class="n">pathway_scores</span> <span class="o">=</span> <span class="n">pathway_scores</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
    
    <span class="n">tfin</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sambar runtime: &quot;</span><span class="p">,</span><span class="n">tfin</span><span class="o">-</span><span class="n">tinit</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">mt</span><span class="p">,</span><span class="n">pathway_scores</span></div>

<div class="viewcode-block" id="binomial_dist"><a class="viewcode-back" href="../../../functions/api.html#netZooPy.sambar.binomial_dist">[docs]</a><span class="k">def</span> <span class="nf">binomial_dist</span><span class="p">(</span><span class="n">u</span><span class="p">,</span><span class="n">v</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Implementation of the binomial dissimilarity funcion or Millar distance from the vegan:vegdist package in R.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">        u : First vector to compare distance from.</span>
<span class="sd">        v : Second vector to compare distance to.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">        bd: _</span>
<span class="sd">            binomial dissimilarity.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">np</span><span class="o">.</span><span class="n">seterr</span><span class="p">(</span><span class="n">divide</span><span class="o">=</span><span class="s1">&#39;ignore&#39;</span><span class="p">,</span> <span class="n">invalid</span><span class="o">=</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span> <span class="c1">#The following steps raise divide by zero errors. This avoids the error to show up as it is not a problem since it is processed afterwards.</span>
    <span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="o">=</span> <span class="n">u</span><span class="p">,</span><span class="n">v</span>
    <span class="n">nk</span> <span class="o">=</span> <span class="n">x</span><span class="o">+</span><span class="n">y</span>
    <span class="n">lognk</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">nk</span><span class="p">))</span>
    <span class="n">t1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">((</span><span class="n">x</span><span class="o">*</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">-</span><span class="n">lognk</span><span class="p">)))</span>
    <span class="n">t2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">((</span><span class="n">y</span><span class="o">*</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">y</span><span class="p">)</span><span class="o">-</span><span class="n">lognk</span><span class="p">)))</span>
    <span class="n">dn</span> <span class="o">=</span> <span class="n">t1</span><span class="o">+</span><span class="n">t2</span><span class="o">+</span><span class="n">nk</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">d</span>  <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">((</span><span class="n">dn</span><span class="o">/</span><span class="n">nk</span><span class="p">)))</span>
    <span class="n">bd</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">d</span><span class="o">&lt;</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">d</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">bd</span></div>

<div class="viewcode-block" id="clustering"><a class="viewcode-back" href="../../../functions/api.html#netZooPy.sambar.clustering">[docs]</a><span class="k">def</span> <span class="nf">clustering</span><span class="p">(</span><span class="n">pt</span><span class="p">,</span> <span class="n">kmin</span><span class="p">,</span><span class="n">kmax</span><span class="p">,</span><span class="n">distance</span><span class="p">,</span><span class="n">linkagem</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Computes the clustering for the pathways matrix and returns a dataframe with the groups with k clusters from kmin to kmax.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">        pt      : Pathway scores.</span>
<span class="sd">        kmin    : Min number of groups in the clustering.</span>
<span class="sd">        kmax    : Max number of groups in the clustering.</span>
<span class="sd">        distance: Similarity metric for the clustering. Default is binomial distance but any distance from scipy.spatial.distance can be used.</span>
<span class="sd">        linkagem: Linkage method. Default is complete.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">        df: _</span>
<span class="sd">            Cluster assignement dataframe.    </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">tinit</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="k">if</span><span class="p">(</span><span class="n">distance</span><span class="o">==</span><span class="s2">&quot;binomial&quot;</span><span class="p">):</span>
        <span class="n">Y</span> <span class="o">=</span> <span class="n">pdist</span><span class="p">(</span><span class="n">pt</span><span class="o">.</span><span class="n">transpose</span><span class="p">(),</span><span class="n">binomial_dist</span><span class="p">)</span> <span class="c1"># Computes the distance matrix. pt is transposed because the pdist function takes rows as input and what we want to cluster are the samples.</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">Y</span> <span class="o">=</span> <span class="n">pdist</span><span class="p">(</span><span class="n">pt</span><span class="o">.</span><span class="n">transpose</span><span class="p">(),</span><span class="n">distance</span><span class="p">)</span>
    <span class="n">Z</span> <span class="o">=</span> <span class="n">linkage</span><span class="p">(</span><span class="n">Y</span><span class="p">,</span><span class="n">linkagem</span><span class="p">)</span> <span class="c1">#Linkage of the clusters using the distance matrix and the complete method.</span>
    <span class="n">np</span><span class="o">.</span><span class="n">savetxt</span><span class="p">(</span><span class="s2">&quot;dist_matrix.csv&quot;</span><span class="p">,</span><span class="n">squareform</span><span class="p">(</span><span class="n">Y</span><span class="p">),</span><span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">)</span><span class="c1">#Saves the distance matrix. Note that the output of pdist is a condensed matrix!!</span>

    <span class="c1">#Building the output dataframe.</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">kmin</span><span class="p">,</span><span class="n">kmax</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">R</span> <span class="o">=</span> <span class="n">cut_tree</span><span class="p">(</span><span class="n">Z</span><span class="p">,</span><span class="n">k</span><span class="p">)</span> <span class="c1">#Cuts the tree at k groups.</span>
        <span class="n">u</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">R</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">k</span><span class="p">)]</span> <span class="o">=</span><span class="n">u</span>
        <span class="n">df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">columns</span>
    <span class="n">tfin</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Clustering runtime: &quot;</span><span class="p">,</span><span class="n">tfin</span><span class="o">-</span><span class="n">tinit</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">df</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span></div>

<div class="viewcode-block" id="sambar"><a class="viewcode-back" href="../../../functions/api.html#netZooPy.sambar.sambar">[docs]</a><span class="k">def</span> <span class="nf">sambar</span><span class="p">(</span><span class="n">mut_file</span><span class="o">=</span><span class="n">mut</span><span class="p">,</span><span class="n">esize_file</span><span class="o">=</span><span class="n">esize</span><span class="p">,</span><span class="n">genes_file</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span><span class="n">gmtfile</span><span class="o">=</span><span class="n">sign</span><span class="p">,</span><span class="n">normPatient</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">kmin</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="n">kmax</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span><span class="n">gmtMSigDB</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">subcangenes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">distance</span><span class="o">=</span><span class="s2">&quot;binomial&quot;</span><span class="p">,</span><span class="n">linkagem</span><span class="o">=</span><span class="s2">&quot;complete&quot;</span><span class="p">,</span><span class="n">cluster</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Runs SAMBAR and outputs the pt matrix, the mt matrix and the clustering matrix.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ------------------</span>
<span class="sd">        mutfile    : matrix of mutations with genes as columns and samples as rows. Format CSV.</span>
<span class="sd">        esize_file : file with genes and their length.</span>
<span class="sd">        genes_file : file with list of cancer-associated genes.</span>
<span class="sd">        gmt_file   : genelist by pathway, format from MSigDB.</span>
<span class="sd">        normPatien : Normalize mutation data by number of mutations in a sample.</span>
<span class="sd">        kmin,kmax  : Number of groups in the clustering.</span>
<span class="sd">        gmtMSigDB  : Whether the signature file comes from MSigDB or not. (Important for processing the file).</span>
<span class="sd">        subcangenes: Makes optional subsetting to cancer-associated genes.</span>
<span class="sd">        distance   : Similarity metric for the clustering. Default is binomial distance but any distance from scipy.spatial.distance can be used.</span>
<span class="sd">        linkagem   : Linkage method. Default is complete.</span>
<span class="sd">        cluster    : Whether the clustering has to be compute or the output will just be the pathway mutation scores.</span>
<span class="sd">    </span>
<span class="sd">    Returns	</span>
<span class="sd">    -------</span>
<span class="sd">        pt :_</span>
<span class="sd">             the pathway matrix.</span>
<span class="sd">        groups :_</span>
<span class="sd">            the groups dataframe as python objects.</span>
<span class="sd">        mt_out.csv :_</span>
<span class="sd">            processed gene mutation scores.</span>
<span class="sd">        pt_out.csv:_ </span>
<span class="sd">            pathway mutation scores</span>
<span class="sd">        clustergroups.csv: _</span>
<span class="sd">            matrix of pertinence to a group in the clustering.</span>
<span class="sd">        dist_matrix.csv  : _</span>
<span class="sd">            Computation of the distance matrix is resource-consuming so the matrix is writen so it doesn&#39;t have to be computed again.</span>

<span class="sd">    Reference:</span>
<span class="sd">        Kuijjer, Marieke Lydia, et al. &quot;Cancer subtype identification using somatic mutation data.&quot; British journal of cancer 118.11 (2018): 1492-1501.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mut</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">mut_file</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">esize</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">esize_file</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">genes_file</span><span class="p">)</span> <span class="c1"># Loads the cancer associated gene list file</span>
    <span class="n">gene_line</span> <span class="o">=</span> <span class="n">file</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">gene_line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">))</span>    
    <span class="n">mt</span><span class="p">,</span><span class="n">pt</span> <span class="o">=</span> <span class="n">desparsify</span><span class="p">(</span><span class="n">mut</span><span class="p">,</span><span class="n">esize</span><span class="p">,</span><span class="n">gmtfile</span><span class="p">,</span><span class="n">genes</span><span class="p">,</span><span class="n">normPatient</span><span class="p">,</span><span class="n">gmtMSigDB</span><span class="p">,</span><span class="n">subcangenes</span><span class="p">)</span>
    
    <span class="n">pt</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;pt_out.csv&quot;</span><span class="p">)</span>
    <span class="n">mt</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;mt_out.csv&quot;</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="p">(</span><span class="ow">not</span> <span class="n">cluster</span><span class="p">):</span> <span class="k">return</span> <span class="n">pt</span>
    
    <span class="n">groups</span> <span class="o">=</span> <span class="n">clustering</span><span class="p">(</span><span class="n">pt</span><span class="p">,</span><span class="n">kmin</span><span class="p">,</span><span class="n">kmax</span><span class="p">,</span><span class="n">distance</span><span class="p">,</span><span class="n">linkagem</span><span class="p">)</span>
    <span class="n">groups</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;clustergroups.csv&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">pt</span><span class="p">,</span><span class="n">groups</span></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2019, netZoo.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>  

  <style>
    /* Sidebar header (and topbar for mobile) */
    /*.wy-side-nav-search, .wy-nav-top {
      background: #fff0;
      sidebartextcolor : #000000;
    }*/
    /* Sidebar */
    /*.wy-nav-side {
     background: #ffffff;
    }*/

    .wy-side-nav-search .wy-dropdown > a img.logo, .wy-side-nav-search > a img.logo {
    width: 100pt;
}
  </style>


</body>
</html>