<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Command Line Interface &mdash; netZooPy 0.9.1 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="The Animals (API)" href="api.html" />
    <link rel="prev" title="Installation guide" href="../install/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> netZooPy
            <img src="../_static/netZoo_logo_v14.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                0.9
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../install/index.html">Installation guide</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Command Line Interface</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#commands">Commands</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#netzoopy">netzoopy</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#netzoopy-panda">panda</a></li>
<li class="toctree-l4"><a class="reference internal" href="#netzoopy-lioness">lioness</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#run-functions-legacy">Run functions (legacy)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#panda">Panda</a></li>
<li class="toctree-l3"><a class="reference internal" href="#lioness">Lioness</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="api.html">The Animals (API)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Home</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">Functions</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutos/index.html">Tutorials</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">netZooPy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Command Line Interface</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/functions/cli.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="command-line-interface">
<h1>Command Line Interface<a class="headerlink" href="#command-line-interface" title="Permalink to this headline"></a></h1>
<p>Netzoopy also has a command line interface that allows to run the animal methods
from the command line, for example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">netzoopy</span> <span class="n">panda</span> <span class="o">--</span><span class="n">e</span> <span class="n">expression</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">m</span> <span class="n">motif</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">p</span> <span class="n">ppi</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">o</span> <span class="n">output_panda</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>The command above would run the panda message passing method using the expression,
motif and ppi data passed as input, and returning an output file containing the
panda weighted edges.</p>
<p>While we expect to provide most animals from CLI, so far you can use: panda,lioness.</p>
<p>If you want to use the previous command line calls (python run_panda.py), check the
<a class="reference internal" href="#run-functions-legacy"><span class="std std-ref">Run functions (legacy)</span></a></p>
<section id="commands">
<h2>Commands<a class="headerlink" href="#commands" title="Permalink to this headline"></a></h2>
<section id="netzoopy">
<h3>netzoopy<a class="headerlink" href="#netzoopy" title="Permalink to this headline"></a></h3>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>netzoopy <span class="o">[</span>OPTIONS<span class="o">]</span> COMMAND <span class="o">[</span>ARGS<span class="o">]</span>...
</pre></div>
</div>
<section id="netzoopy-panda">
<h4>panda<a class="headerlink" href="#netzoopy-panda" title="Permalink to this headline"></a></h4>
<p>Run panda using expression, motif and ppi data.
Use flags to modify the function behavior. By default, boolean flags are false.
Output is a text file, with the TF, Gene, Motif, Force columns, where TF and Gene
are the nodes of the network considered, Motif is the prior and force is
the actual panda score.</p>
<p>Example:</p>
<blockquote>
<div><p>netzoopy panda -e tests/puma/ToyData/ToyExpressionData.txt -m tests/puma/ToyData/ToyMotifData.txt -p tests/puma/ToyData/ToyPPIData.txt -o test_panda.txt</p>
</div></blockquote>
<p>Reference:</p>
<blockquote>
<div><p>Glass, Kimberly, et al. “Passing messages between biological networks to refine predicted interactions.” PloS one 8.5 (2013): e64832.</p>
</div></blockquote>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>netzoopy panda <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-panda-e">
<span id="cmdoption-netzoopy-panda-expression"></span><span class="sig-name descname"><span class="pre">-e</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--expression</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;expression&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-panda-e" title="Permalink to this definition"></a></dt>
<dd><p><strong>Required</strong> Path to file containing the gene expression data. By default,                   the expression file does not have a header, and the cells are separated by a tab.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-panda-m">
<span id="cmdoption-netzoopy-panda-motif"></span><span class="sig-name descname"><span class="pre">-m</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--motif</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;motif&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-panda-m" title="Permalink to this definition"></a></dt>
<dd><p><strong>Required</strong> Path to pair file containing the transcription factor DNA binding motif edges in the form of TF-gene-weight(0/1). If not provided, the gene coexpression matrix is returned as a result network.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-panda-p">
<span id="cmdoption-netzoopy-panda-ppi"></span><span class="sig-name descname"><span class="pre">-p</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--ppi</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;ppi&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-panda-p" title="Permalink to this definition"></a></dt>
<dd><p><strong>Required</strong> Path to pair file containing the PPI edges. The PPI can be symmetrical, if not, it will be transformed into a symmetrical adjacency matrix.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-panda-o">
<span id="cmdoption-netzoopy-panda-out"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--out</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-panda-o" title="Permalink to this definition"></a></dt>
<dd><p><strong>Required</strong> Output panda file. Format as txt</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-panda-computing">
<span class="sig-name descname"><span class="pre">--computing</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;computing&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-panda-computing" title="Permalink to this definition"></a></dt>
<dd><p>computing option, choose one between cpu and gpu</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>cpu</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-panda-precision">
<span class="sig-name descname"><span class="pre">--precision</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;precision&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-panda-precision" title="Permalink to this definition"></a></dt>
<dd><p>precision option</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>double</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-panda-save_memory">
<span id="cmdoption-netzoopy-panda-save-memory"></span><span class="sig-name descname"><span class="pre">--save_memory</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-netzoopy-panda-save_memory" title="Permalink to this definition"></a></dt>
<dd><p>panda option. When true the result network is weighted adjacency matrix of size (nTFs, nGenes).                  when false The result network has 4 columns in the form gene - TF - weight in motif prior - PANDA edge.</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>False</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-panda-save_tmp">
<span id="cmdoption-netzoopy-panda-save-tmp"></span><span class="sig-name descname"><span class="pre">--save_tmp</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-netzoopy-panda-save_tmp" title="Permalink to this definition"></a></dt>
<dd><p>panda option</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>False</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-panda-rm_missing">
<span id="cmdoption-netzoopy-panda-rm-missing"></span><span class="sig-name descname"><span class="pre">--rm_missing</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-netzoopy-panda-rm_missing" title="Permalink to this definition"></a></dt>
<dd><p>Removes the genes and TFs that are not present in one of the priors. Works only if modeProcess=legacy</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-panda-keep_expr">
<span id="cmdoption-netzoopy-panda-keep-expr"></span><span class="sig-name descname"><span class="pre">--keep_expr</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-netzoopy-panda-keep_expr" title="Permalink to this definition"></a></dt>
<dd><p>Keeps the input expression matrix in the result</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>False</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-panda-mode_process">
<span id="cmdoption-netzoopy-panda-mode-process"></span><span class="sig-name descname"><span class="pre">--mode_process</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;mode_process&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-panda-mode_process" title="Permalink to this definition"></a></dt>
<dd><p>panda option for input data processing. Choose between union(default),                   legacy and intersection</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>union</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-panda-alpha">
<span class="sig-name descname"><span class="pre">--alpha</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;alpha&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-panda-alpha" title="Permalink to this definition"></a></dt>
<dd><p>panda and lioness first sample</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>0.1</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-panda-start">
<span class="sig-name descname"><span class="pre">--start</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;start&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-panda-start" title="Permalink to this definition"></a></dt>
<dd><p>panda and lioness first sample</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>1</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-panda-end">
<span class="sig-name descname"><span class="pre">--end</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;end&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-panda-end" title="Permalink to this definition"></a></dt>
<dd><p>panda and lioness last sample</p>
</dd></dl>

</section>
<section id="netzoopy-lioness">
<h4>lioness<a class="headerlink" href="#netzoopy-lioness" title="Permalink to this headline"></a></h4>
<p>Run Lioness to extract single-sample networks.
First runs panda using expression, motif and ppi data.
Then runs lioness and puts results in the output_lioness folder.
Use flags to modify the function behavior. By default, boolean flags are false.</p>
<p>Example:</p>
<blockquote>
<div><p>netzoopy lioness -e tests/puma/ToyData/ToyExpressionData.txt -m tests/puma/ToyData/ToyMotifData.txt -p tests/puma/ToyData/ToyPPIData.txt -op test_panda.txt -ol lioness/</p>
</div></blockquote>
<dl class="simple">
<dt>Reference:</dt><dd><p>Kuijjer, Marieke Lydia, et al. “Estimating sample-specific regulatory networks.” Iscience 14 (2019): 226-240.</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>netzoopy lioness <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-e">
<span id="cmdoption-netzoopy-lioness-expression"></span><span class="sig-name descname"><span class="pre">-e</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--expression</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;expression&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-e" title="Permalink to this definition"></a></dt>
<dd><p><strong>Required</strong> Path to file containing the gene expression data. By default,                   the expression file does not have a header, and the cells are separated by a tab.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-m">
<span id="cmdoption-netzoopy-lioness-motif"></span><span class="sig-name descname"><span class="pre">-m</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--motif</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;motif&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-m" title="Permalink to this definition"></a></dt>
<dd><p><strong>Required</strong> Path to pair file containing the transcription factor DNA binding motif edges in the form of TF-gene-weight(0/1). If not provided, the gene coexpression matrix is returned as a result network.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-p">
<span id="cmdoption-netzoopy-lioness-ppi"></span><span class="sig-name descname"><span class="pre">-p</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--ppi</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;ppi&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-p" title="Permalink to this definition"></a></dt>
<dd><p><strong>Required</strong> Path to pair file containing the PPI edges. The PPI can be symmetrical, if not, it will be transformed into a symmetrical adjacency matrix.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-op">
<span id="cmdoption-netzoopy-lioness-out-panda"></span><span class="sig-name descname"><span class="pre">-op</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--out-panda</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output_panda&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-op" title="Permalink to this definition"></a></dt>
<dd><p><strong>Required</strong> Output panda file. Format as txt</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-ol">
<span id="cmdoption-netzoopy-lioness-out-lioness"></span><span class="sig-name descname"><span class="pre">-ol</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--out-lioness</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output_lioness&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-ol" title="Permalink to this definition"></a></dt>
<dd><p><strong>Required</strong> Output lioness folder</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-fmt">
<span class="sig-name descname"><span class="pre">--fmt</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;fmt&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-fmt" title="Permalink to this definition"></a></dt>
<dd><p>Lioness network files output format. Choose one between .npy,.txt,.mat</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>npy</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-computing">
<span class="sig-name descname"><span class="pre">--computing</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;computing&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-computing" title="Permalink to this definition"></a></dt>
<dd><p>computing option, choose one between cpu and gpu</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>cpu</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-precision">
<span class="sig-name descname"><span class="pre">--precision</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;precision&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-precision" title="Permalink to this definition"></a></dt>
<dd><p>precision option</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>double</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-ncores">
<span class="sig-name descname"><span class="pre">--ncores</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;ncores&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-ncores" title="Permalink to this definition"></a></dt>
<dd><p>Number of cores. Lioness CPU parallelizes over ncores</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>1</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-save_memory">
<span id="cmdoption-netzoopy-lioness-save-memory"></span><span class="sig-name descname"><span class="pre">--save_memory</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-save_memory" title="Permalink to this definition"></a></dt>
<dd><p>panda option. When true the result network is weighted adjacency matrix of size (nTFs, nGenes).                  when false The result network has 4 columns in the form gene - TF - weight in motif prior - PANDA edge.</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>False</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-save_tmp">
<span id="cmdoption-netzoopy-lioness-save-tmp"></span><span class="sig-name descname"><span class="pre">--save_tmp</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-save_tmp" title="Permalink to this definition"></a></dt>
<dd><p>panda option</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>False</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-rm_missing">
<span id="cmdoption-netzoopy-lioness-rm-missing"></span><span class="sig-name descname"><span class="pre">--rm_missing</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-rm_missing" title="Permalink to this definition"></a></dt>
<dd><p>Removes the genes and TFs that are not present in one of the priors. Works only if modeProcess=legacy</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-mode_process">
<span id="cmdoption-netzoopy-lioness-mode-process"></span><span class="sig-name descname"><span class="pre">--mode_process</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;mode_process&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-mode_process" title="Permalink to this definition"></a></dt>
<dd><p>panda option for input data processing. Choose between union(default),                   legacy and intersection</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>union</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-output_type">
<span id="cmdoption-netzoopy-lioness-output-type"></span><span class="sig-name descname"><span class="pre">--output_type</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output_type&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-output_type" title="Permalink to this definition"></a></dt>
<dd><p>lioness option for output format. Choose one between network, gene_targeting, tf_targeting</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>network</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-alpha">
<span class="sig-name descname"><span class="pre">--alpha</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;alpha&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-alpha" title="Permalink to this definition"></a></dt>
<dd><p>panda and lioness first sample</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>0.1</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-start">
<span class="sig-name descname"><span class="pre">--start</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;start&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-start" title="Permalink to this definition"></a></dt>
<dd><p>panda and lioness first sample</p>
<dl class="field-list simple">
<dt class="field-odd">Default</dt>
<dd class="field-odd"><p>1</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-netzoopy-lioness-end">
<span class="sig-name descname"><span class="pre">--end</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;end&gt;</span></span><a class="headerlink" href="#cmdoption-netzoopy-lioness-end" title="Permalink to this definition"></a></dt>
<dd><p>panda and lioness last sample</p>
</dd></dl>

</section>
</section>
</section>
<section id="run-functions-legacy">
<h2>Run functions (legacy)<a class="headerlink" href="#run-functions-legacy" title="Permalink to this headline"></a></h2>
<section id="panda">
<h3>Panda<a class="headerlink" href="#panda" title="Permalink to this headline"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="netZooPy.panda.run_panda.main">
<span class="sig-prename descclassname"><span class="pre">netZooPy.panda.run_panda.</span></span><span class="sig-name descname"><span class="pre">main</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">argv</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/netZooPy/panda/run_panda.html#main"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#netZooPy.panda.run_panda.main" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Description:</dt><dd><p>Run PANDA algorithm from the command line.</p>
</dd>
<dt>Inputs:</dt><dd><p>-h, –help: help
-e, –expression: expression_file : Path to file containing the gene expression data. By default, the expression file does not have a header, and the cells ares separated by a tab.
-m, –motif: Path to pair file containing the transcription factor DNA binding motif edges in the form of TF-gene-weight(0/1). If not provided, the gene coexpression matrix is returned as a result network.
-p, –ppi: Path to pair file containing the PPI edges. The PPI can be symmetrical, if not, it will be transformed into a symmetrical adjacency matrix.
-o, –out: output file.
-r, –rm_missing.
-q, –lioness: output for Lioness single sample networks.</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>python run_panda.py -e ../../tests/ToyData/ToyExpressionData.txt -m ../../tests/ToyData/ToyMotifData.txt -p ../../tests/ToyData/ToyPPIData.txt -o test_panda.txt -q output_panda.txt</p>
<dl class="simple">
<dt>Reference:</dt><dd><p>Glass, Kimberly, et al. “Passing messages between biological networks to refine predicted interactions.” PloS one 8.5 (2013): e64832.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="lioness">
<h3>Lioness<a class="headerlink" href="#lioness" title="Permalink to this headline"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="netZooPy.lioness.run_lioness.main">
<span class="sig-prename descclassname"><span class="pre">netZooPy.lioness.run_lioness.</span></span><span class="sig-name descname"><span class="pre">main</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">argv</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/netZooPy/lioness/run_lioness.html#main"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#netZooPy.lioness.run_lioness.main" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Description:</dt><dd><p>Run LIONESS algorithm from the command line.</p>
</dd>
<dt>Usage:</dt><dd><p>-h, –help: help
-e, –expression: expression matrix (.npy)
-m, –motif: motif matrix, normalized (.npy)
-p, –ppi: ppi matrix, normalized (.npy)
-g, –comp: use cpu (default) or gpu
-r, –pre: number of digits to calcluate
-c, –ncores: number cores
-n, –npy: PANDA network (.npy)
-o, –out: output folder
-f, –format: output format (txt, npy, or mat)
start: to start from nth sample (optional)
end: to end at nth sample (optional, must with start)</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>python3 run_lioness.py -e ../../tests/ToyData/ToyExpressionData.txt -m ../../tests/ToyData/ToyMotifData.txt -p ../../tests/ToyData/ToyPPIData.txt -g cpu -r single -c 2 -o /tmp -f npy 1 2</p>
<dl class="simple">
<dt>Reference:</dt><dd><p>Kuijjer, Marieke Lydia, et al. “Estimating sample-specific regulatory networks.” Iscience 14 (2019): 226-240.</p>
</dd>
</dl>
</dd></dl>

</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../install/index.html" class="btn btn-neutral float-left" title="Installation guide" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="api.html" class="btn btn-neutral float-right" title="The Animals (API)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2019, netZoo.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>  

  <style>
    /* Sidebar header (and topbar for mobile) */
    /*.wy-side-nav-search, .wy-nav-top {
      background: #fff0;
      sidebartextcolor : #000000;
    }*/
    /* Sidebar */
    /*.wy-nav-side {
     background: #ffffff;
    }*/

    .wy-side-nav-search .wy-dropdown > a img.logo, .wy-side-nav-search > a img.logo {
    width: 100pt;
}
  </style>


</body>
</html>