<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>netZooPy.puma.puma &mdash; netZooPy 0.9.1 documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../../index.html" class="icon icon-home"> netZooPy
            <img src="../../../_static/netZoo_logo_v14.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                0.9
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../install/index.html">Installation guide</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../functions/cli.html">Command Line Interface</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../functions/api.html">The Animals (API)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">Home</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../functions/index.html">Functions</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../tutos/index.html">Tutorials</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">netZooPy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../../index.html">Module code</a> &raquo;</li>
      <li>netZooPy.puma.puma</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for netZooPy.puma.puma</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">print_function</span>

<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">.timer</span> <span class="kn">import</span> <span class="n">Timer</span>
<span class="kn">from</span> <span class="nn">netZooPy.panda.panda</span> <span class="kn">import</span> <span class="n">Panda</span>
<span class="kn">import</span> <span class="nn">netZooPy.panda.calculations</span> <span class="k">as</span> <span class="nn">calc</span>
<span class="kn">from</span> <span class="nn">netZooPy.puma.calculations</span> <span class="kn">import</span> <span class="n">compute_puma</span>


<div class="viewcode-block" id="Puma"><a class="viewcode-back" href="../../../functions/api.html#netZooPy.puma.puma.Puma">[docs]</a><span class="k">class</span> <span class="nc">Puma</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Description:</span>
<span class="sd">        Using PUMA to infer gene regulatory network.</span>
<span class="sd">        1. Reading in input data (expression data, motif prior, TF PPI data, miR)</span>
<span class="sd">        2. Computing coexpression network</span>
<span class="sd">        3. Normalizing networks</span>
<span class="sd">        4. Running PUMA algorithm</span>
<span class="sd">        5. Writing out PUMA network</span>

<span class="sd">    Inputs:</span>
<span class="sd">        object: Puma object.</span>

<span class="sd">     Methods:</span>
<span class="sd">        __init__                    : Intialize instance of Puma class.</span>
<span class="sd">        __remove_missing            : Removes the gens and TFs that are not present in one of the priors. Works only if modeProcess=&#39;legacy&#39;.</span>
<span class="sd">        _normalize_network          : Standardizes the input data matrices.</span>
<span class="sd">        puma_loop                   : The PUMA algorithm.</span>
<span class="sd">        __pearson_results_data_frame: Saves PUMA network in edges format.</span>
<span class="sd">        save_puma_results           : Saves PUMA network.</span>
<span class="sd">        top_network_plot            : Selects top genes to plot.</span>
<span class="sd">        __shape_plot_network        : Creates network plot.</span>
<span class="sd">        __create_plot               : Runs network plot.</span>
<span class="sd">        return_puma_indegree        : computes indegree of puma network, only if save_memory = False.</span>
<span class="sd">        return_puma_outdegree       : computes outdegree of puma network, only if save_memory = False.</span>

<span class="sd">    Example:</span>
<span class="sd">        Run the PUMA algorithm, leave out motif and PPI data to use Pearson correlation network:</span>
<span class="sd">        from netZooPy.puma.puma import Puma</span>
<span class="sd">        puma_obj = Puma(&#39;../../tests/ToyData/ToyExpressionData.txt&#39;, &#39;../../tests/ToyData/ToyMotifData.txt&#39;, &#39;../../tests/ToyData/ToyPPIData.txt&#39;,&#39;../../tests/ToyData/ToyMiRList.txt&#39;)</span>

<span class="sd">     Authors:</span>
<span class="sd">       cychen, davidvi, alessandromarin</span>

<span class="sd">    Reference:</span>
<span class="sd">        Kuijjer, Marieke L., et al. &quot;PUMA: PANDA Using MicroRNA Associations.&quot; BioRxiv (2019).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">expression_file</span><span class="p">,</span>
        <span class="n">motif_file</span><span class="p">,</span>
        <span class="n">ppi_file</span><span class="p">,</span>
        <span class="n">mir_file</span><span class="p">,</span>
        <span class="n">modeProcess</span><span class="o">=</span><span class="s2">&quot;union&quot;</span><span class="p">,</span>
        <span class="n">computing</span><span class="o">=</span><span class="s2">&quot;cpu&quot;</span><span class="p">,</span>
        <span class="n">precision</span><span class="o">=</span><span class="s2">&quot;double&quot;</span><span class="p">,</span>
        <span class="n">save_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">save_tmp</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">remove_missing</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">keep_expression_matrix</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">alpha</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
        <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">end</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Description:</span>
<span class="sd">            Intialize instance of Puma class and load data.</span>

<span class="sd">        Inputs:</span>
<span class="sd">            expression_file : Path to file containing the gene expression data.</span>
<span class="sd">            motif_file      : Path to file containing the regulation prior. This can be a miRNA-Gene predicted network from TargetScan/miRanda.</span>
<span class="sd">                              However, this can be combined with transcription factor DNA binding motif data in the form of TF-gene-weight(0/1) to estimate gene regulation by TF and miRNA.</span>
<span class="sd">                              If set to none, the gene coexpression matrix is returned as a result network.</span>
<span class="sd">            ppi_file        : Path to file containing the TF PPI data. This can be provided as &#39;None&#39; if no TF data is given and PUMA will estimate a miRNA-Gene networks.</span>
<span class="sd">            mir_file        : Path to file containing miRNA list.</span>
<span class="sd">            computing       : &#39;cpu&#39; uses Central Processing Unit (CPU) to run PANDA.</span>
<span class="sd">                              &#39;gpu&#39; use the Graphical Processing Unit (GPU) to run PANDA.</span>
<span class="sd">            precision       : &#39;double&#39; computes the regulatory network in double precision (15 decimal digits).</span>
<span class="sd">                              &#39;single&#39; computes the regulatory network in single precision (7 decimal digits) which is fastaer, requires half the memory but less accurate.</span>

<span class="sd">            save_memory     : True : removes temporary results from memory. The result network is weighted adjacency matrix of size (nTFs, nGenes).</span>
<span class="sd">                              False: keeps the temporary files in memory. The result network has 4 columns in the form gene - TF - weight in motif prior - PUMA edge.</span>
<span class="sd">            save_tmp        : Save temporary variables.</span>
<span class="sd">            remove_missing  : Removes the gens and TFs that are not present in one of the priors. Works only if modeProcess=&#39;legacy&#39;.</span>
<span class="sd">            keep_expression_matrix: Keeps the input expression matrix in the result Puma object.</span>
<span class="sd">            modeProcess     : The input data processing mode.</span>
<span class="sd">                              &#39;legacy&#39;: refers to the processing mode in netZooPy&lt;=0.5</span>
<span class="sd">                              (Default)&#39;union&#39;: takes the union of all TFs and genes across priors and fills the missing genes in the priors with zeros.</span>
<span class="sd">                              &#39;intersection&#39;: intersects the input genes and TFs across priors and removes the missing TFs/genes.</span>
<span class="sd">            alpha           : Learning rate (default: 0.1)</span>
<span class="sd">            start           : first sample of expression (default 1)</span>
<span class="sd">            end             : last sample of expression (default None)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># =====================================================================</span>
        <span class="c1"># Data loading</span>
        <span class="c1"># =====================================================================</span>
        <span class="n">Panda</span><span class="o">.</span><span class="n">processData</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">modeProcess</span><span class="p">,</span>
            <span class="n">motif_file</span><span class="p">,</span>
            <span class="n">expression_file</span><span class="p">,</span>
            <span class="n">ppi_file</span><span class="p">,</span>
            <span class="n">remove_missing</span><span class="p">,</span>
            <span class="n">keep_expression_matrix</span><span class="p">,</span>
            <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
            <span class="n">end</span><span class="o">=</span><span class="n">end</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">with</span> <span class="n">Timer</span><span class="p">(</span><span class="s2">&quot;Loading miR data ...&quot;</span><span class="p">):</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">mir_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">miR</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>
            <span class="n">TFNames</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">unique_tfs</span>
            <span class="n">sort_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">TFNames</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">s1</span> <span class="o">=</span> <span class="n">sort_idx</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="n">TFNames</span><span class="p">,</span> <span class="n">miR</span><span class="p">,</span> <span class="n">sorter</span><span class="o">=</span><span class="n">sort_idx</span><span class="p">)]</span>

        <span class="k">if</span> <span class="n">remove_missing</span> <span class="ow">and</span> <span class="n">motif_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">__remove_missing</span><span class="p">()</span>

        <span class="c1"># =====================================================================</span>
        <span class="c1"># Network construction</span>
        <span class="c1"># =====================================================================</span>
        <span class="k">with</span> <span class="n">Timer</span><span class="p">(</span><span class="s2">&quot;Calculating coexpression network ...&quot;</span><span class="p">):</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression_data</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">correlation_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">identity</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">num_genes</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">correlation_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">expression_data</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">correlation_matrix</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
                <span class="n">np</span><span class="o">.</span><span class="n">fill_diagonal</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">correlation_matrix</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">correlation_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">correlation_matrix</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_data</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="s2">&quot;Returning the correlation matrix of expression data in &lt;Puma_obj&gt;.correlation_matrix&quot;</span>
            <span class="p">)</span>
            <span class="c1"># self.puma_network = self.correlation_matrix</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">__pearson_results_data_frame</span><span class="p">()</span>
            <span class="k">return</span>
        <span class="c1"># Auxiliary dicts</span>
        <span class="n">gene2idx</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_names</span><span class="p">)}</span>
        <span class="n">tf2idx</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">unique_tfs</span><span class="p">)}</span>

        <span class="k">with</span> <span class="n">Timer</span><span class="p">(</span><span class="s2">&quot;Creating motif network ...&quot;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">motif_matrix_unnormalized</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">num_tfs</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_genes</span><span class="p">))</span>
            <span class="n">idx_tfs</span> <span class="o">=</span> <span class="p">[</span><span class="n">tf2idx</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_data</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
            <span class="n">idx_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene2idx</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_data</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
            <span class="n">idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ravel_multi_index</span><span class="p">(</span>
                <span class="p">(</span><span class="n">idx_tfs</span><span class="p">,</span> <span class="n">idx_genes</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_matrix_unnormalized</span><span class="o">.</span><span class="n">shape</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">motif_matrix_unnormalized</span><span class="o">.</span><span class="n">ravel</span><span class="p">()[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_data</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ppi_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">identity</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">num_tfs</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">Timer</span><span class="p">(</span><span class="s2">&quot;Creating PPI network ...&quot;</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">ppi_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">identity</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">num_tfs</span><span class="p">)</span>
                <span class="n">idx_tf1</span> <span class="o">=</span> <span class="p">[</span><span class="n">tf2idx</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
                <span class="n">idx_tf2</span> <span class="o">=</span> <span class="p">[</span><span class="n">tf2idx</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
                <span class="n">idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ravel_multi_index</span><span class="p">((</span><span class="n">idx_tf1</span><span class="p">,</span> <span class="n">idx_tf2</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">ppi_matrix</span><span class="o">.</span><span class="n">ravel</span><span class="p">()[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
                <span class="n">idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ravel_multi_index</span><span class="p">((</span><span class="n">idx_tf2</span><span class="p">,</span> <span class="n">idx_tf1</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">ppi_matrix</span><span class="o">.</span><span class="n">ravel</span><span class="p">()[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>

        <span class="c1"># =====================================================================</span>
        <span class="c1"># Network normalization</span>
        <span class="c1"># =====================================================================</span>
        <span class="k">with</span> <span class="n">Timer</span><span class="p">(</span><span class="s2">&quot;Normalizing networks ...&quot;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">correlation_matrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_network</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">correlation_matrix</span><span class="p">)</span>
            <span class="k">with</span> <span class="n">np</span><span class="o">.</span><span class="n">errstate</span><span class="p">(</span><span class="n">invalid</span><span class="o">=</span><span class="s2">&quot;ignore&quot;</span><span class="p">):</span>  <span class="c1"># silly warning bothering people</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">motif_matrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_network</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">motif_matrix_unnormalized</span>
                <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ppi_matrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_network</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ppi_matrix</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">precision</span> <span class="o">==</span> <span class="s2">&quot;single&quot;</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">correlation_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">correlation_matrix</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">motif_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_matrix</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">ppi_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ppi_matrix</span><span class="p">)</span>

        <span class="c1"># =====================================================================</span>
        <span class="c1"># Clean up useless variables to release memory</span>
        <span class="c1"># =====================================================================</span>
        <span class="k">if</span> <span class="n">save_memory</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Clearing motif and ppi data, unique tfs, and gene names for speed&quot;</span><span class="p">)</span>
            <span class="k">del</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">motif_data</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">unique_tfs</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">gene_names</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">motif_matrix_unnormalized</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="c1"># =====================================================================</span>
        <span class="c1"># Saving middle data to tmp</span>
        <span class="c1"># =====================================================================</span>
        <span class="k">if</span> <span class="n">save_tmp</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">Timer</span><span class="p">(</span><span class="s2">&quot;Saving expression matrix and normalized networks ...&quot;</span><span class="p">):</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;/tmp/expression.npy&quot;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression_data</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
                <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;/tmp/motif.normalized.npy&quot;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_matrix</span><span class="p">)</span>
                <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;/tmp/ppi.normalized.npy&quot;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_matrix</span><span class="p">)</span>

        <span class="c1"># Clean up useless variables to release memory</span>
        <span class="k">if</span> <span class="n">keep_expression_matrix</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">expression_matrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression_data</span><span class="o">.</span><span class="n">values</span>
        <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression_data</span>

        <span class="c1"># =====================================================================</span>
        <span class="c1"># Running PUMA algorithm</span>
        <span class="c1"># =====================================================================</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Running PUMA algorithm ...&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">puma_network</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">puma_loop</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">correlation_matrix</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">motif_matrix</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ppi_matrix</span><span class="p">,</span>
            <span class="n">computing</span><span class="p">,</span>
            <span class="n">alpha</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">__remove_missing</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Description:</span>
<span class="sd">            Removes the gens and TFs that are not present in one of the priors. Works only if modeProcess=&#39;legacy&#39;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Remove expression not in motif:&quot;</span><span class="p">)</span>
            <span class="n">motif_unique_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_data</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">len_tot</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">expression_data</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">expression_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression_data</span><span class="p">[</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">expression_data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">motif_unique_genes</span><span class="p">)</span>
            <span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene_names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression_data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">num_genes</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_names</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="s2">&quot;   </span><span class="si">{}</span><span class="s2"> rows removed from the initial </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">len_tot</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_genes</span><span class="p">,</span> <span class="n">len_tot</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="c1"># if self.motif_data is not None:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Remove motif not in expression data:&quot;</span><span class="p">)</span>
        <span class="n">len_tot</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">motif_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_data</span><span class="p">[</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">motif_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_names</span><span class="p">)</span>
        <span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">unique_tfs</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_data</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_tfs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">unique_tfs</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="s2">&quot;   </span><span class="si">{}</span><span class="s2"> rows removed from the initial </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">len_tot</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_data</span><span class="p">),</span> <span class="n">len_tot</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Remove ppi not in motif:&quot;</span><span class="p">)</span>
            <span class="n">motif_unique_tfs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">])</span>
            <span class="n">len_tot</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">[</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">motif_unique_tfs</span><span class="p">)</span>
            <span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">[</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">motif_unique_tfs</span><span class="p">)</span>
            <span class="p">]</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="s2">&quot;   </span><span class="si">{}</span><span class="s2"> rows removed from the initial </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">len_tot</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">),</span> <span class="n">len_tot</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">_normalize_network</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Description:</span>
<span class="sd">            Standardizes the input data matrices.</span>

<span class="sd">        Inputs:</span>
<span class="sd">            x     : Input adjacency matrix.</span>

<span class="sd">        Outputs:</span>
<span class="sd">            normalized_matrix: Standardized adjacency matrix.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">calc</span><span class="o">.</span><span class="n">normalize_network</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">puma_loop</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">correlation_matrix</span><span class="p">,</span> <span class="n">motif_matrix</span><span class="p">,</span> <span class="n">ppi_matrix</span><span class="p">,</span> <span class="n">computing</span><span class="o">=</span><span class="s2">&quot;cpu&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.1</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Description:</span>
<span class="sd">            The PUMA algorithm.</span>

<span class="sd">        Inputs:</span>
<span class="sd">            correlation_matrix: Input coexpression matrix.</span>
<span class="sd">            motif_matrix      : Input motif regulation prior network.</span>
<span class="sd">            ppi_matrix        : Input PPI matrix.</span>
<span class="sd">            computing         : &#39;cpu&#39; uses Central Processing Unit (CPU) to run PANDA.</span>
<span class="sd">                                &#39;gpu&#39; use the Graphical Processing Unit (GPU) to run PANDA.</span>

<span class="sd">        Methods:</span>
<span class="sd">            t_function      : Continuous Tanimoto similarity function computed on the CPU.</span>
<span class="sd">            update_diagonal : Updates the diagonal of the input matrix in the message passing computed on the CPU.</span>
<span class="sd">            gt_function     : Continuous Tanimoto similarity function computed on the GPU.</span>
<span class="sd">            gupdate_diagonal: Updates the diagonal of the input matrix in the message passing computed on the GPU.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">puma_loop_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
        <span class="c1"># TODO:This should be using self.correlation. Keeping for retrocompatibility</span>
        <span class="n">motif_matrix</span> <span class="o">=</span> <span class="n">compute_puma</span><span class="p">(</span>
            <span class="n">correlation_matrix</span><span class="p">,</span>
            <span class="n">motif_matrix</span><span class="p">,</span>
            <span class="n">ppi_matrix</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">s1</span><span class="p">,</span>
            <span class="n">computing</span><span class="o">=</span><span class="n">computing</span><span class="p">,</span>
            <span class="n">alpha</span><span class="o">=</span><span class="n">alpha</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Running puma took: </span><span class="si">%.2f</span><span class="s2"> seconds!&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">puma_loop_time</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">puma_network</span> <span class="o">=</span> <span class="n">motif_matrix</span>
        <span class="c1"># Ale: reintroducing the export_puma_results array if Puma called with save_memory=False</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;unique_tfs&quot;</span><span class="p">):</span>
            <span class="n">tfs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">unique_tfs</span><span class="p">,</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_names</span><span class="p">),</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
            <span class="n">genes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_names</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_tfs</span><span class="p">)</span>
            <span class="n">motif</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_matrix_unnormalized</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span><span class="n">order</span><span class="o">=</span><span class="s2">&quot;F&quot;</span><span class="p">)</span>
            <span class="n">force</span> <span class="o">=</span> <span class="n">motif_matrix</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span><span class="n">order</span><span class="o">=</span><span class="s2">&quot;F&quot;</span><span class="p">)</span>
            <span class="c1"># TODO: should we keep formats consistent between Panda and Puma?</span>
            <span class="c1"># self.export_puma_results = pd.DataFrame({&#39;tf&#39;:tfs, &#39;gene&#39;: genes,&#39;motif&#39;: motif, &#39;force&#39;: force})</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">export_puma_results</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">column_stack</span><span class="p">((</span><span class="n">tfs</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">motif</span><span class="p">,</span> <span class="n">force</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">motif_matrix</span>

    <span class="k">def</span> <span class="nf">__pearson_results_data_frame</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Description:</span>
<span class="sd">            Saves PUMA network in edges format.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">genes_1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_names</span><span class="p">,</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_names</span><span class="p">),</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="n">genes_2</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_names</span><span class="p">,</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_names</span><span class="p">),</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">flat_puma_network</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">puma_network</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">export_puma_results</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="p">{</span><span class="s2">&quot;tf&quot;</span><span class="p">:</span> <span class="n">genes_1</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="n">genes_2</span><span class="p">,</span> <span class="s2">&quot;force&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">flat_puma_network</span><span class="p">}</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">export_puma_results</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">export_puma_results</span><span class="p">[[</span><span class="s2">&quot;tf&quot;</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;force&quot;</span><span class="p">]]</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">save_puma_results</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="s2">&quot;puma.npy&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Description:</span>
<span class="sd">            Saves PUMA network.</span>

<span class="sd">        Inputs:</span>
<span class="sd">            path: Path to save the network.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="n">Timer</span><span class="p">(</span><span class="s2">&quot;Saving PUMA network to </span><span class="si">%s</span><span class="s2"> ...&quot;</span> <span class="o">%</span> <span class="n">path</span><span class="p">):</span>
            <span class="c1"># Because there are two modes of operation (save_memory), save to file will be different</span>
            <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;export_puma_results&quot;</span><span class="p">):</span>
                <span class="n">toexport</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">export_puma_results</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">toexport</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">puma_network</span>
            <span class="c1"># Export to file</span>
            <span class="k">if</span> <span class="n">path</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.txt&quot;</span><span class="p">):</span>
                <span class="n">np</span><span class="o">.</span><span class="n">savetxt</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">toexport</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot; &quot;</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">path</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.csv&quot;</span><span class="p">):</span>
                <span class="n">np</span><span class="o">.</span><span class="n">savetxt</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">toexport</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">path</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.tsv&quot;</span><span class="p">):</span>
                <span class="n">np</span><span class="o">.</span><span class="n">savetxt</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">toexport</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">toexport</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">top_network_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s2">&quot;puma_top_100.png&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Description:</span>
<span class="sd">            Selects top genes.</span>

<span class="sd">        Inputs:</span>
<span class="sd">            top        : Top number of genes to plot.</span>
<span class="sd">            file       : File to save the network plot.</span>
<span class="sd">            plot_bipart: Plot the network as a bipartite layout.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;export_puma_results&quot;</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">AttributeError</span><span class="p">(</span>
                <span class="s2">&quot;Puma object does not contain the export_puma_results attribute.</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="o">+</span> <span class="s2">&quot;Run Puma with the flag save_memory=False&quot;</span>
            <span class="p">)</span>
        <span class="c1"># Ale TODO: work in numpy instead of pandas?</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">puma_results</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">export_puma_results</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;tf&quot;</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;motif&quot;</span><span class="p">,</span> <span class="s2">&quot;force&quot;</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">subset_puma_results</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">puma_results</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span>
            <span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;force&quot;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span>
        <span class="p">)</span>
        <span class="n">subset_puma_results</span> <span class="o">=</span> <span class="n">subset_puma_results</span><span class="p">[</span>
            <span class="n">subset_puma_results</span><span class="o">.</span><span class="n">tf</span> <span class="o">!=</span> <span class="n">subset_puma_results</span><span class="o">.</span><span class="n">gene</span>
        <span class="p">]</span>
        <span class="n">subset_puma_results</span> <span class="o">=</span> <span class="n">subset_puma_results</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">top</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__shape_plot_network</span><span class="p">(</span><span class="n">subset_puma_results</span><span class="o">=</span><span class="n">subset_puma_results</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">file</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">__shape_plot_network</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">subset_puma_results</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s2">&quot;puma.png&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Description:</span>
<span class="sd">            Creates plot.</span>

<span class="sd">        Inputs:</span>
<span class="sd">            subset_puma_results : Reduced PUMA network to the top genes.</span>
<span class="sd">            file                : File to save the network plot.</span>
<span class="sd">            plot_bipart         : Plot the network as a bipartite layout.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># reshape data for networkx</span>
        <span class="n">unique_genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
            <span class="nb">set</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">subset_puma_results</span><span class="p">[</span><span class="s2">&quot;tf&quot;</span><span class="p">])</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">subset_puma_results</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]))</span>
        <span class="p">)</span>
        <span class="n">unique_genes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">)</span>
        <span class="n">unique_genes</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
        <span class="n">unique_genes</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">unique_genes</span><span class="o">.</span><span class="n">index</span>
        <span class="n">subset_puma_results</span> <span class="o">=</span> <span class="n">subset_puma_results</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="n">unique_genes</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="s2">&quot;tf&quot;</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;name&quot;</span>
        <span class="p">)</span>
        <span class="n">subset_puma_results</span> <span class="o">=</span> <span class="n">subset_puma_results</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="s2">&quot;tf_index&quot;</span><span class="p">})</span>
        <span class="n">subset_puma_results</span> <span class="o">=</span> <span class="n">subset_puma_results</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s2">&quot;name&quot;</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">subset_puma_results</span> <span class="o">=</span> <span class="n">subset_puma_results</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="n">unique_genes</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;name&quot;</span>
        <span class="p">)</span>
        <span class="n">subset_puma_results</span> <span class="o">=</span> <span class="n">subset_puma_results</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
            <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="s2">&quot;gene_index&quot;</span><span class="p">}</span>
        <span class="p">)</span>
        <span class="n">subset_puma_results</span> <span class="o">=</span> <span class="n">subset_puma_results</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s2">&quot;name&quot;</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">links</span> <span class="o">=</span> <span class="n">subset_puma_results</span><span class="p">[[</span><span class="s2">&quot;tf_index&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_index&quot;</span><span class="p">,</span> <span class="s2">&quot;force&quot;</span><span class="p">]]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__create_plot</span><span class="p">(</span><span class="n">unique_genes</span><span class="o">=</span><span class="n">unique_genes</span><span class="p">,</span> <span class="n">links</span><span class="o">=</span><span class="n">links</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">file</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">__create_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">unique_genes</span><span class="p">,</span> <span class="n">links</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s2">&quot;puma.png&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Description:</span>
<span class="sd">            Runs the plot.</span>

<span class="sd">        Inputs:</span>
<span class="sd">            unique_genes : Unique list of PUMA genes.</span>
<span class="sd">            links        : Edges of the subset PUMA network to the top genes.</span>
<span class="sd">            file         : File to save the network plot.</span>
<span class="sd">            plot_bipart  : Plot the network as a bipartite layout.</span>

<span class="sd">        Methods:</span>
<span class="sd">            split_label: Splits the plot label over several lines for plotting purposes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>
        <span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

        <span class="n">g</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
        <span class="n">g</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">clf</span><span class="p">()</span>
        <span class="n">g</span><span class="o">.</span><span class="n">add_nodes_from</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">])</span>
        <span class="n">edges</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">links</span><span class="p">)):</span>
            <span class="n">edges</span> <span class="o">=</span> <span class="n">edges</span> <span class="o">+</span> <span class="p">[</span>
                <span class="p">(</span>
                    <span class="n">links</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;tf_index&quot;</span><span class="p">],</span>
                    <span class="n">links</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;gene_index&quot;</span><span class="p">],</span>
                    <span class="nb">float</span><span class="p">(</span><span class="n">links</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;force&quot;</span><span class="p">])</span> <span class="o">/</span> <span class="mi">200</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">]</span>
        <span class="n">g</span><span class="o">.</span><span class="n">add_weighted_edges_from</span><span class="p">(</span><span class="n">edges</span><span class="p">)</span>
        <span class="n">labels</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">def</span> <span class="nf">split_label</span><span class="p">(</span><span class="n">label</span><span class="p">):</span>
            <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Description: Splits the plot label over several lines for plotting purposes.</span>

<span class="sd">            Inputs:</span>
<span class="sd">                label: Input label text.</span>

<span class="sd">            Outputs:</span>
<span class="sd">                label: Output label text divided over several lines.</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="n">ll</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">label</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">ll</span> <span class="o">&gt;</span> <span class="mi">6</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">label</span><span class="p">[</span><span class="mi">0</span> <span class="p">:</span> <span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">ll</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)]</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">label</span><span class="p">[</span><span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">ll</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span> <span class="p">:]</span>
            <span class="k">return</span> <span class="n">label</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_genes</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]):</span>
            <span class="n">labels</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">split_label</span><span class="p">(</span><span class="n">l</span><span class="p">)</span>
        <span class="n">pos</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">spring_layout</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>
        <span class="c1"># nx.draw_networkx(g, pos, labels=labels, node_size=40, font_size=3, alpha=0.3, linewidths = 0.5, width =0.5)</span>
        <span class="n">colors</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">edges</span><span class="p">))</span>
        <span class="n">options</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;alpha&quot;</span><span class="p">:</span> <span class="mf">0.7</span><span class="p">,</span>
            <span class="s2">&quot;edge_color&quot;</span><span class="p">:</span> <span class="n">colors</span><span class="p">,</span>
            <span class="s2">&quot;edge_cmap&quot;</span><span class="p">:</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">Blues</span><span class="p">,</span>
            <span class="s2">&quot;node_size&quot;</span><span class="p">:</span> <span class="mi">110</span><span class="p">,</span>
            <span class="s2">&quot;vmin&quot;</span><span class="p">:</span> <span class="o">-</span><span class="mi">100</span><span class="p">,</span>
            <span class="s2">&quot;width&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
            <span class="s2">&quot;labels&quot;</span><span class="p">:</span> <span class="n">labels</span><span class="p">,</span>
            <span class="s2">&quot;font_weight&quot;</span><span class="p">:</span> <span class="s2">&quot;regular&quot;</span><span class="p">,</span>
            <span class="s2">&quot;font_size&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;linewidths&quot;</span><span class="p">:</span> <span class="mi">20</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="n">nx</span><span class="o">.</span><span class="n">draw_spring</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span> <span class="n">iterations</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="o">**</span><span class="n">options</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;off&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">return_puma_indegree</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Description:</span>
<span class="sd">            computes indegree of PUMA network, only if save_memory = False.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># subset_indegree = self.export_puma_results.loc[:,[&#39;gene&#39;,&#39;force&#39;]]</span>
        <span class="n">subset_indegree</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">puma_results</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;force&quot;</span><span class="p">]]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">puma_indegree</span> <span class="o">=</span> <span class="n">subset_indegree</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">puma_indegree</span>

    <span class="k">def</span> <span class="nf">return_puma_outdegree</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Description:</span>
<span class="sd">            computes outdegree of PUMA network, only if save_memory = False.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">export_puma_results_pd</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">export_puma_results</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;tf&quot;</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;motif&quot;</span><span class="p">,</span> <span class="s2">&quot;force&quot;</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">subset_outdegree</span> <span class="o">=</span> <span class="n">export_puma_results_pd</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="s2">&quot;tf&quot;</span><span class="p">,</span> <span class="s2">&quot;force&quot;</span><span class="p">]]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">puma_outdegree</span> <span class="o">=</span> <span class="n">subset_outdegree</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;tf&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">puma_outdegree</span></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2019, netZoo.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>  

  <style>
    /* Sidebar header (and topbar for mobile) */
    /*.wy-side-nav-search, .wy-nav-top {
      background: #fff0;
      sidebartextcolor : #000000;
    }*/
    /* Sidebar */
    /*.wy-nav-side {
     background: #ffffff;
    }*/

    .wy-side-nav-search .wy-dropdown > a img.logo, .wy-side-nav-search > a img.logo {
    width: 100pt;
}
  </style>


</body>
</html>